README.md
# number to english translation (1 to 1000)

This a simple tool for learning purposes; it simply accepts command line entries of numbers between 1 to 1000 from the user and prints out a grammatically correct english translation of the number. 

## Requirements
Make sure you have JDK 8+ installed on your machine. Project also requires leiningen build tool so download it from https://leiningen.org/. 

## Installation & Usage
Download via git clone and then run `lein run` at the command line. Run `lein test` to run tests.

Alternatively you can run `lein uberjar` and this will create a standalone jar to be distributed. This can be run with the following command: `java -jar target/uberjar/num--0.1.0-SNAPSHOT-standalone.jar`.

When the tool is started, user can follow the directions on screen.

## License


This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2019-02-16
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2019-02-16
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/num-/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/num-/compare/0.1.0...0.1.1


test/num_/core_test.clj
(ns num-.core-test
  (:require [clojure.test :refer :all]
            [num-.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


test/numtoenglish/core_test.clj
(ns numtoenglish.core-test
  (:require [clojure.test :refer :all]
            [numtoenglish.core :refer :all]))

(deftest number->english->tests
  (testing "1 should return one"
    (is (= "one" (number->english 1))))
  (testing "241 should return two hundred and forty one"
    (is (= "two hundred and forty one" (number->english 241))))
  (testing "1000 should return one thousand"
    (is (= "one thousand" (number->english 1000)))))

(deftest boundary-tests
  (testing "non numeric input should return false"
    (is (= false (check-boundary "some_name")))))

(deftest only-tens-conversion
  (testing "42 should return forty two"
    (is (= "forty two" (number->english 42)))))

(deftest only-hundreds-conversion
  (testing "200 should return two hundred"
    (is (= "two hundred" (number->english 200)))))

(deftest hundreds->englist-test
  (testing "2 should return as two hundred"
    (is (= "two hundred" (hundreds->english 2)))))

(deftest tens->english-test
  (testing "30 should return as thirty"
    (is (= "thirty" (tens->english 30)))))

(deftest below-hundred-test
  (testing "74 should return seventy four"
    (is (= "seventy four" (below-hundred->english 74)))))

(deftest hundreds->english-test
  (testing "0 should return nil from this function"
    (is (= nil (hundreds->english 0)))))

(deftest isnumeric-test
  (testing "123 should return true"
    (is (= true (is-numeric "123"))))
  (testing "erkin123 should return false"
    (is (= false (is-numeric "erkin123")))))


src/numtoenglish/convert.clj
(ns numtoenglish.convert)

(def small-numbers
  {1 "one",
   2 "two",
   3 "three",
   4 "four",
   5 "five",
   6 "six",
   7 "seven",
   8 "eight",
   9 "nine",
   10 "ten",
   11 "eleven",
   12 "twelve",
   13 "thirteen",
   14 "fourteen",
   15 "fifteen",
   16 "sixteen",
   17 "seventeen",
   18 "eighteen",
   19 "nineteen"})

(def ten-products
  {20 "twenty",
   30 "thirty",
   40 "forty",
   50 "fifty",
   60 "sixty",
   70 "seventy",
   80 "eighty",
   90 "ninety"})

(defn check-boundary
  [n]
  (and (integer? n) (> n 0) (<= n 1000)))

(defn hundreds->english
  [digit]
  (when (not= digit 0)
      (str (get small-numbers digit) " hundred")))

(defn tens->english
  [digit]
  (if (= digit 0) ""
      (str (get ten-products digit))))

(defn below-hundred->english
  [num]
  (if (<= num 20)
    (get small-numbers num)
    (str
     (tens->english (* 10 (int (/ num 10))))
     " "
     (get small-numbers (mod num 10)))))

(defn concat-numbers
  [hunds tens]
  (cond
    (and (some? hunds) (some? tens)) (str hunds " and " tens)
    (some? hunds) hunds
    (some? tens) tens
    :else ""))

(defn number->english
  "converts a small number by converting hundreds and tens separately"
  [number]
  (if (= number 1000)
    (str "one thousand")
    (concat-numbers
     (hundreds->english (int (/ number 100)))
     (below-hundred->english (mod number 100)))))


(defn is-numeric
  "checks if a word supplied is numeric"
  [word]
  (= 0 (count (filter #(not (Character/isDigit %)) word))))

(defn convert
  "this is the main entry point for conversion, does the IO loop"
  []
  (do
    (println "Number to English converter; please enter a number between 1 to 1000 inclusive; to quit press Enter")
    (loop [number (read-line)]
      (when (not= number "")
        (if (is-numeric number) 
          (do
            (let [num (Integer. number)]
              (if (check-boundary num)
                (println (str "English: " (number->english num)))
                (println "Please enter a number between 1 to 1000 inclusive"))))
          (println "Please enter a number between 1 to 1000 inclusive"))
        (recur (read-line)))) ))
